src/cockpit/channels/filesystem.py

```diff
@@ -77,39 +79,35 @@ def do_open(self, options):
             self.close()
 
 
-class FsReadChannel(Channel):
+class FsReadChannel(GeneratorChannel):
     payload = 'fsread1'
 
-    def do_open(self, options):
-        self.ready()
+    def get_data(self, size_hint: int, options: Dict[str, object]) -> Generator[bytes, int, Dict[str, object]]:
+        binary = options.get('binary', False)
+        max_read_size = options.get('max_read_size')
+
+        logger.debug('Opening file "%s" for reading', options['path'])
+
         try:
-            logger.debug('Opening file "%s" for reading', options['path'])
-            try:
-                with open(options['path'], 'rb') as filep:
-                    buf = os.stat(filep.fileno())
-                    tag = tag_from_stat(buf)
-                    if max_read_size := options.get('max_read_size'):
-                        if buf.st_size > max_read_size:
-                            raise ChannelError('too-large')
-
-                    data = filep.read()
-            except FileNotFoundError:
-                self.close(tag='-')
-                return
-            except PermissionError:
-                raise ChannelError('access-denied')
-            except OSError:
-                raise ChannelError('internal-error')
+            with open(options['path'], 'rb') as filep:
+                buf = os.stat(filep.fileno())
+                if max_read_size is not None and buf.st_size > max_read_size:
+                    raise ChannelError('too-large')
 
-            if 'binary' not in options:
-                data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
+                while (data := filep.read(size_hint)):
```

Comment:
Finally, `size_hint` is documented to allow 0 or negative, and that shouldn't just be stuffed into .read{,1}, but needs to be handled separately.

Reply:
Hmm, on second sight this is not true -- that documentation applied to `send_data()`. It also wouldn't make sense to even call get_data() with ≤ 0, because what could it even do with that information.. (sleep? yield None? all of that can be done by the parent class if necessary)

Reply:
... but I still don't understand this. What exactly happens in line 101 when it calls `size_hint  = yield data`? What kind of return values does that get, and when?

Reply:
This is what comes from the `__generator.send()` call in the `GeneratorChannel`.  We get the new size_hint here and will use it to make a read() and then yield the result, which becomes the return value of `.send()` back in the `GeneratorChannel`

Reply:
Note: the reason we do this is slightly non-obvious.  It's not expected that the size hint will ever change, so the generator could keep reading values of the original size.  But: if for some reason we have an incomplete read or the size changes (for example, through our character set conversion for non-binary channels) then we'll have sent a partial block.  We can use the size_hint (ie: up to the remainder of the block size) to get back on track again.

That said: I just notice that this value comes from the return value of `.send_data()`, and the required logic isn't actually implemented there.  Oops.

Reply:
That said, I realized over the weekend that I remember this wrong.  That was the idea of my original approach, but I changed tack, and the new approach actually doesn't depend on this calculation at all.  It only depends on the sender sending "reasonable-sized" chunks, ideally as close to the "ping" size as possible.

Reply:
...which also means that we can skip the complicated theatrics with the bidirectional generator, and ask the generator to always send blocks of the same size.  We could also replace the integer return value of `.send_data()` with a `bool` meaning "is it OK to send more".

Reply:
Did that.